app/src/main/java/io/github/polysmee/database/FirebaseUploadService.java

```diff
@@ -1,40 +1,76 @@
 package io.github.polysmee.database;
 
+import android.content.Context;
+import android.os.Build;
+
 import androidx.annotation.NonNull;
 
 import com.google.firebase.storage.FirebaseStorage;
 import com.google.firebase.storage.StorageReference;
 
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Paths;
+
 import io.github.polysmee.database.databaselisteners.DownloadValueListener;
 import io.github.polysmee.database.databaselisteners.LoadValueListener;
 
+import static java.nio.file.Files.readAllBytes;
+
 public final class FirebaseUploadService implements UploadService {
 
     @Override
-    public void uploadImage(@NonNull byte[] data, String fileName, LoadValueListener onSuccess, LoadValueListener onFailure) {
+    public void uploadImage(@NonNull byte[] data, String fileName, LoadValueListener onSuccess, LoadValueListener onFailure, Context ctx) {
         String imageName = "" + System.currentTimeMillis() + fileName;
+        addNewFileToCache(imageName, data, ctx);
         StorageReference ref = FirebaseStorage.getInstance().getReference().child(imageName);
         ref
-                .putBytes(data)
-                .addOnSuccessListener(taskSnapshot -> onSuccess.onDone(imageName))
-                .addOnFailureListener(ignored -> onFailure.onDone(ignored.getMessage()));
+            .putBytes(data)
+            .addOnSuccessListener(taskSnapshot -> onSuccess.onDone(imageName))
+            .addOnFailureListener(ignored -> onFailure.onDone(ignored.getMessage()));
     }
 
     @Override
-    public void downloadImage(String id, DownloadValueListener dvl, LoadValueListener fl) {
-        StorageReference ref = FirebaseStorage.getInstance().getReference().child(id);
-        ref
+    public void downloadImage(String id, DownloadValueListener dvl, LoadValueListener fl, Context ctx) {
```

Comment:
Isn't it possible to add the new downloaded image to cache as well? Because right now, we only add a image to cache if we're the ones uploading it

Reply:
No, we are looking in the cache using the getFileFromCache method